src/Core/Context/ContextBuilderInitializer.php

```diff
@@ -0,0 +1,57 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Context;
+
+use Context;
+
+/**
+ * This service is used to help initialize multiple context builders, this is convenient
+ * for install processes or in tests, but it shouldn't be used in other use cases.
+ *
+ * Usually the context builders are initialized via request listeners.
+ *
+ * @internal
+ */
+class ContextBuilderInitializer
+{
+    public function __construct(
+        private readonly LanguageContextBuilder $languageContextBuilder,
+    ) {
+    }
+
+    public function initializeFromLegacyContext(Context $context): void
+    {
+        if ($context->language && $context->language->id) {
+            $this->languageContextBuilder->setLanguageId($context->language->id);
+        }
+    }
+
+    public function setLanguageId(int $languageId): void
+    {
+        $this->languageContextBuilder->setLanguageId($languageId);
+    }
+}
```

Comment:
Here we are talking about context initialization, but we only use a setter. The build is done when calling the lazy service factory. So perhaps we are talking about prerequisites rather than initialization?

Reply:
(I'm talking about the we of the class which disturbs me)

Reply:
Also I have the impression that we could directly call the builder, right? This perhaps saves us from duplication

Reply:
Indeed this service only prepares the Builders, not the context themselves It initializes the `*ContextBuilder` services (not all for now but it can evolve).

The reason I didn't want to use the builder directly is that it forces us to set them as public, because these edge cases are places where you can't inject a service you can only fetch some via `Container::get` It means that ultimately all builders would become public piece by piece and I wanted to avoid that at all costs

So the alternative, which is far from perfect I'm sorry, was to have a service that is public and can influence the builders so instead of forcing all builders to be public we can only have this particular initializer service And I also flagged it as internal because it's not supposed to be used by anyone else at all

I almost hesitate to put some kind of safety inside the service methods so that they check what code executed them and throw an exception when it's not somewhere from the core, to prevent modules from using this service 😅 

Now that you have more context maybe you can help me find a better naming because I'm stuck Or maybe you would have a better idea that could meet with these constraints?

Reply:
`ContextBuilderManager`? 😅 `MultipleContextBuilder`, `ContextBuilderPreparator` or at some point I was thinking about simply `ContextBuilder`... nothing stands out that feels perfect

Reply:
ContextBuilderInitializer gives me the impression that it is possible to build contexts, or context builders... it's starting to look weird x) Preparator is good. At least you understand that it's just for setup

Reply:
I renamed it `ContextBuilderPreparer` (preparator is incorrect english actually), and also renamed the methods with `prepare` as a prefix so it should prevent any confusion